### ISSUE TITLE
NRE in new SK2 2.0 networking code

### ISSUE THREAD

**Issue by JustArchi** [SteamRE/SteamKit#400]
```
2017-06-16 15:36:05|mono-sgen-1935|FATAL|ASF|UnobservedTaskExceptionHandler() System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.NullReferenceException: Object reference not set to an instance of an object
  at SteamKit2.TcpConnection.TryConnect (System.Object sender) [0x00094] in <f740b1bfddfa4ddb95d88fffe3231fb2>:0 
  at SteamKit2.TcpConnection+<>c__DisplayClass19_0.<Connect>b__0 (System.Threading.Tasks.Task`1[TResult] t) [0x000b2] in <f740b1bfddfa4ddb95d88fffe3231fb2>:0 
  at System.Threading.Tasks.ContinuationTaskFromResultTask`1[TAntecedentResult].InnerInvoke () [0x00024] in <73ee1b14c2fa4d61b481096ff3d8d6d7>:0 
  at System.Threading.Tasks.Task.Execute () [0x00010] in <73ee1b14c2fa4d61b481096ff3d8d6d7>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object
  at SteamKit2.TcpConnection.TryConnect (System.Object sender) [0x00094] in <f740b1bfddfa4ddb95d88fffe3231fb2>:0 
  at SteamKit2.TcpConnection+<>c__DisplayClass19_0.<Connect>b__0 (System.Threading.Tasks.Task`1[TResult] t) [0x000b2] in <f740b1bfddfa4ddb95d88fffe3231fb2>:0 
  at System.Threading.Tasks.ContinuationTaskFromResultTask`1[TAntecedentResult].InnerInvoke () [0x00024] in <73ee1b14c2fa4d61b481096ff3d8d6d7>:0 
  at System.Threading.Tasks.Task.Execute () [0x00010] in <73ee1b14c2fa4d61b481096ff3d8d6d7>:0 <---
```

This looks similar to https://github.com/SteamRE/SteamKit/issues/383 but stack trace is a bit different, it looks like the issue happens when there is no connection of any kind between given machine and Steam, but there was connection before. Refer to https://github.com/JustArchi/ArchiSteamFarm/issues/576 for original issue. @Netshroud this is probably related to new .NET core code.

Thank you in advance for looking into this. Let me know if you need debug log.

**Comment by JustArchi** [comment]
One more to the collection:

```
2017-07-21 02:51:53|dotnet-19313|FATAL|ASF|OnUnhandledException() System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.Sockets.Socket'.
   at System.Net.Sockets.Socket.Poll(Int32 microSeconds, SelectMode mode)
   at SteamKit2.TcpConnection.NetLoop()
   at System.Threading.Thread.ThreadMain_ThreadStart()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
```

**Comment by JustArchi** [comment]
And one more:

```
2017-08-15 21:02:51|dotnet-31362|FATAL|ASF|OnUnhandledException() System.NullReferenceException: Object reference not set to an instance of an object.
   at SteamKit2.TcpConnection.TryConnect(Object sender)
   at SteamKit2.TcpConnection.Connect(EndPoint endPoint, Int32 timeout)
   at SteamKit2.Internal.CMClient.<>c__DisplayClass45_0.<Connect>b__0(Task`1 t)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at SteamKit2.Internal.CMClient.Disconnect()
   at SteamKit2.Internal.CMClient.Connect(ServerRecord cmServer)
   at ArchiSteamFarm.Bot.<Connect>d__105.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ArchiSteamFarm.Bot.<OnDisconnected>d__138.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch() 
```

**Comment by JustArchi** [comment]
Reproduced also with UDP:

```
2017-08-22 19:33:50|ArchiSteamFarm-10268|FATAL|ASF|OnUnhandledException() System.NullReferenceException: Object reference not set to an instance of an object.
   at SteamKit2.UdpConnection.SendPacket(UdpPacket packet)
   at SteamKit2.UdpConnection.SendPendingMessages()
   at SteamKit2.UdpConnection.NetLoop(Object param)
   at System.Threading.Thread.ThreadMain_ParameterizedThreadStart(Object parameter)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state) 
```

**Comment by azuisleet** [comment]
I plan to review the changes in 2.0 for concurrency issues sometime soon. Our last discussion ended on a discussion about the logic being hoisted from TCPConnection into CMClient. I suspect that we do indeed need locking inside CMClient ("connectLock" from TCPConnection), but also the refactoring from #402 still needs to be merged as well.

**Comment by JustArchi** [comment]
Don't mind me, I'm just reporting anything new that I find unusual so you can take a closer look and decide what (and if) you want to do anything with that. Thanks!

**Comment by yaakov-h** [comment]
Hopefully this isn't possible anymore with the new networking code.

 As usual, feel free to re-open or create a new issue if it happens again.

**Comment by JustArchi** [comment]
Waiting patiently for new SK2 release, I'll be happy to test once it's ready 👍
